pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/RuntimeASTTransformer.groovy

```diff
@@ -1238,10 +1249,37 @@ class RuntimeASTTransformer {
             // These variable handles are declared in functions, but will still be bound to the script context
             // Doing this here ensures the variables make the trip across to run time - if they don't make it, neither did this pipeline
             pipelineBlock.addStatement(stmt(callX(closureX(currentBlock), 'call')))
-            pipelineElementHandles.clear()
         }
 
-        /**
+        @NonNull
+        private List<Statement> prepareClosureScopedHandles(@NonNull BlockStatement pipelineBlock) {
+            ArrayList<Statement> result = new ArrayList<Statement>()
+            if (SCRIPT_SPLITTING_TRANSFORMATION) {
+                ArrayList<DeclarationExpression> declarations = new ArrayList<DeclarationExpression>()
+                moduleNode.statementBlock.statements.each { item ->
+                    if (((ExpressionStatement) item).expression instanceof DeclarationExpression) {
+                        declarations.add((DeclarationExpression) ((ExpressionStatement) item).expression)
+                    }
+                }
+
+                if (declarations.size() != 0) {
+                    if (SCRIPT_SPLITTING_ALLOW_LOCAL_VARIABLES) {
+                        result.addAll(pipelineElementHandles)
+                        pipelineElementHandles.clear()
+                    } else {
+                        throw new IllegalStateException("[JENKINS-34987] SCRIPT_SPLITTING_TRANSFORMATION is an experimental feature of Declarative Pipeline and is incompatible with local variable declarations inside a Jenkinsfile. " +
+                                "Add the '@Field' annotation to local variable declarations: " +
+                                declarations.join(", ") + ". To use less effective script splitting which allows local variable declarations without changing your pipeline code, set SCRIPT_SPLITTING_ALLOW_LOCAL_VARIABLES=true .")
```

Comment:
```suggestion
                        throw new IllegalStateException("[JENKINS-34987] SCRIPT_SPLITTING_TRANSFORMATION is an experimental feature of Declarative Pipeline and is incompatible with local variable declarations inside a Jenkinsfile. " +
                                "Local variable declarations found: " +
                                declarations.join(", ") + ". Use of Groovy var 
                                 "As a temporary workaround, you can add the '@Field' annotation to these local variable declarations. " +
                                 "However, use of Groovy variables in Declarative pipeline, with or without '@Field' annotation, is not recommended or supported. " + 
                                 "To use less effective script splitting which allows local variable declarations without changing your pipeline code, set SCRIPT_SPLITTING_ALLOW_LOCAL_VARIABLES=true .")
```

Reply:
@bitwiseman 
Is this definition `SCRIPT_SPLITTING_ALLOW_LOCAL_VARIABLES=true` will have permanent affect or it is also implemented as temporary workaround?

Reply:
@Moskovych
I'm not sure I understand your question.  The setting will have effect for as long as you keep it.  If you turn it off, pipelines that were broken before will be broken again.   

The question of whether this setting will be there in future versions is not decided yet.  Script splitting as a whole is still experimental and may change.

Reply:
Let's assume I didn't enabled anything. So means by default it's disabled.
Just updated plugins to 1.8.2. So why the pipelines broken then?